### ISSUE TITLE
[Discussion] Where are the mappings curated from?

### ISSUE THREAD

**Issue by parkerlreed** [schellingb/dosbox-pure#41]
Is this a local database that has every game? Are they downloaded on demand?

Would love to know more about this.

**Comment by schellingb** [comment]
The mappings as of now are all embedded in the core dll file. We'll see what a future version brings in terms of modifying the mappings in some way without the need of updating the entire core.

The [keyb2joypad.cpp](../blob/main/keyb2joypad.cpp) file contains all the mappings and button and game names in compressed form. When a game is loaded, before the start menu shows up, the files are evaluated to find a known key identification file which decides which mapping to use. It uses a pre-filled hash table so it is nice and quick.

So, as mentioned in the options and the source, the "keyb2joy.pad" (you'll find a bit more information if you google that, with the . and the quotes) project is the source for the mappings. There is a google sheets document (which I think is editable so I can't share the link) with 3000 or so rows, one per game and lots of columns for all buttons and notes, etc.

I made a script which converts that document to the database as used by DOSBox Pure. I also made a script that parses several known DOS game file databases (almost 3 million files) to find these identifying files in a manner that tries to work around some games existing in many patched or localized versions. For example, there are 20 variants of "DOOM.EXE" in the file list, so it makes more sense to identify DOOM by some other file (although DOOM, and many others, ended up with multiple identifiers, but much less than 20). These scripts are not yet public but hopefully will be once I get around cleaning them up.

If you want to know more, the most detailed description of the project I know can be found on a comment by one of the authors JemyM on one of the pre-release videos I posted onto Youtube. Link to the comment: https://www.youtube.com/watch?v=V4-bcncgyHo&lc=Ugw7491N1QDXa3yD4RJ4AaABAg
